src/main/java/com/google/sps/Recommend.java

```diff
@@ -16,34 +27,134 @@
 
 public class Recommend {
 
+  // keep and use up to this many Interaction entities
+  private static final int UPPER_LIMIT = 5_000_000;
+  // multipliers for score calculation
+  private static final double NO_INTERACTION = 1.5;
+  private static final double ALREADY_VIEWED = 1.2;
+  private static final double ALREADY_SAVED = 0.6;
+
+  // comparator to sort doubles in descending order
+  private static final Comparator<Double> SCORE_DESCENDING =
+      new Comparator<Double>() {
+        @Override
+        public int compare(Double a, Double b) {
+          return Double.compare(b, a);
+        }
+      };
+
+  private static final Logger LOGGER = Logger.getLogger(Recommend.class.getName());
+
   private static SparkSession spark;
   private static DatastoreService datastore;
 
   /** Initializes the SparkSession. */
   private static void init() {
-    spark =
-        SparkSession.builder()
-            .appName("Java Spark SQL basic example")
-            .config("spark.master", "local[*]")
-            .getOrCreate();
-    spark.sparkContext().setLogLevel("ERROR");
-
-    datastore = DatastoreServiceFactory.getDatastoreService();
+    if (spark == null) {
+      spark =
+          SparkSession.builder()
+              .appName("Java Spark SQL basic example")
+              .config("spark.master", "local[*]")
```

Comment:
Not in this PR, but I would recommend you to try a distributed pipeline as well.